### ISSUE TITLE
getBodyStructure wrong with multipart messages

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5763]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)| @pitrou reported|
|-|-|
|Trac ID|trac#5763|
|Type|defect|
|Created|2012-07-10 17:14:34Z|
|Branch|https://github.com/twisted/twisted/tree/imap4-multipart-bodystructure-5763-2|

Judging by [RFC 3501](http://tools.ietf.org/html/rfc3501#page-74) and some [real life examples](http://mailman2.u.washington.edu/pipermail/imap-protocol/2011-May/001413.html), imap4.getBodyStructure() doesn't return a correct value for multipart messages.

Attachments:

* [getbodystructuremultipart.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/419/419d0e49e375f35f49f009174f87788e84f3ce8f/a3ca5f397b18604e00b9b06ebd981e2eb8d99018.patch) (5753 bytes) - added by antoine on 2012-07-10 17:15:05Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5763 5763
type__defect defect
reporter__antoine antoine
priority__normal normal
milestone__ 
branch__branches_imap4_multipart_bodystructure_5763_2 branches/imap4-multipart-bodystructure-5763-2
branch_author__exarkun__antoine exarkun, antoine
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1341940474000000 1341940474000000
changetime__1343051601000000 1343051601000000
version__None None
owner__exarkun exarkun
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_1"></a>|@pitrou commented|
|-|-|

Attached patch claims to fix this, and adds a test.

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|Alex commented|
|-|-|

Hi Antoine,

I'm far from an IMAP expert, but a few notes from trying to review this patch:

* There is a comment at the top of getBodyStructure that indicates it doesn't handle multipart messages correctly, it should be removed with this patch.
* The code in the `True` branch of `if disp:` (both times) is untested.

I've left the review keyword since I don't really feel qualified to do this review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Thanks!  I think you have correctly divined the intent of the IMAP4 specification.  The new behavior seems like an improvement over the old behavior.  `getBodyStructure` is still fairly obtuse, though (not your fault; my fault, I wrote it).  I'm going to have a try at cleaning it up so that it is slightly easier to understand (maybe I can even make it easier to understand than the RFC!).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [34778]) Branching to 'imap4-multipart-bodystructure-5763'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [34779]) Apply getbodystructuremultipart.patch 

refs [#5763](https://github.com/twisted/twisted/issues/5763)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [34786]) Well, maybe this is close (except for, like, docs, and maybe some more tests).  Just two more failing tests (for multipart messages, haha)

refs [#5763](https://github.com/twisted/twisted/issues/5763)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Found some more bugs, sadly.  There's more tests now, which is nice at least.  Some of them don't even run a full loopback IMAP4 client/server pair to exercise the BODYSTRUCTURE code.

I think the BODYSTRUCTURE computing code is clearly now too.  Though I can imagine making it even clearer, and exposing the message structure objects as part of the public _client_ API someday, I think maybe I've done enough in this branch for now.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/imap4-multipart-bodystructure-5763)

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_9"></a>|@pitrou commented|
|-|-|

Certain content-disposition parameters (the date parameters, and any arbitrary parameter) have their values quoted as per RFC 2183:

```
     disposition-parm := filename-parm
                       / creation-date-parm
                       / modification-date-parm
                       / read-date-parm
                       / size-parm
                       / parameter

     filename-parm := "filename" "=" value

     creation-date-parm := "creation-date" "=" quoted-date-time

     modification-date-parm := "modification-date" "=" quoted-date-time

     read-date-parm := "read-date" "=" quoted-date-time

     size-parm := "size" "=" 1*DIGIT
```

Where RFC 2045 defines `parameter` as:

```
     parameter := attribute "=" value

     attribute := token
                  ; Matching of attributes
                  ; is ALWAYS case-insensitive.

     value := token / quoted-string
```

That said, the original code didn't cater for that either, so I'm mainly mentioning this for reference.

I notice you sort the results in _unquotedAttrs, which should fix the issue with hash randomization. Of course, it also means the original order is not respected, which doesn't sound very much of a problem.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph set owner to @exarkun|
|-|-|

Okay.  I think this is ready to land.  I wish I could be surer, but that would require a PhD in IMAP which is beyond the scope of this review.

Some thoughts, though:

  1. It sounds like the issue antoine reported in his last comment should be reported as a separate bug.  Or bugs, if the ordering thing is really important; it seems to me that the previous order is irrelevant there but I might be missing something.
  2. Basically _all_ the new docstrings are invalid epytext.  Please re-format them appropriately, and preview them by generating the documentation locally before merging.  I don't understand why the buildbot isn't telling us about this; I guess that needs to be investigated.  Unfortunately epytext doesn't support block quotes, you will have to use preformatted blocks, although bulleted or numbered lists should be fine for many of the uses.
  3. `getBodyStructure` _is_ a lot more comprehensible, but, uh, still not super comprehensible.  The new, expanded docstring really helps with understanding the return type though.  I still don't entirely understand how this is _possible_ without using more stuff from the 'email' module, though... I guess that, implicitly, any _implementation_ of `IMessagePart` would either have to use it or provide its own whole MIME library?

Thanks very much to the authors.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

> It sounds like the issue antoine reported in his last comment should be reported as a separate bug. Or bugs, if the ordering thing is really important; it seems to me that the previous order is irrelevant there but I might be missing something.

I filed a ticket for [ticket:5799 handling quoting].  The order is irrelevant as far as I know.

> Basically all the new docstrings are invalid epytext. Please re-format them appropriately, and preview them by generating the documentation locally before merging. I don't understand why the buildbot isn't telling us about this; I guess that needs to be investigated. Unfortunately epytext doesn't support block quotes, you will have to use preformatted blocks, although bulleted or numbered lists should be fine for many of the uses.

Fixed those.

> getBodyStructure is a lot more comprehensible, but, uh, still not super comprehensible. The new, expanded docstring really helps with understanding the return type though. I still don't entirely understand how this is possible without using more stuff from the 'email' module, though... I guess that, implicitly, any implementation of IMessagePart would either have to use it or provide its own whole MIME library?

Right.  Although perhaps the email module will offer some header parsing functionality (I haven't checked whether it can handle, eg, Content-Disposition correctly, though).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [34929]) Branching to 'imap4-multipart-bodystructure-5763-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [34933]) Merge imap4-multipart-bodystructure-5763-2

Author: antoine, exarkun
Reviewer: antoine, glyph
Fixes: [#5763](https://github.com/twisted/twisted/issues/5763)

Refactor the IMAP4 server's BODYSTRUCTURE response generation code, adding documentation,
unit tests, and fixing several bugs:

  * incorrect ordering of content-language/content-location
  * incorrect structure for header parameters in multiple places
  * overall incorrect structure for multipart/* messages
  * incorrect empty lists which should instead be NILs
  * failure to ever fill in certain other fields like content-id
